### ISSUE TITLE
Compositional Attention Networks for Machine Reasoning

### ISSUE THREAD

**Issue by takoroy** [takoroy/ml-study-memo#3]
# 概要
論理的推論のための新しいネットワークアーキテクチャであるMAC Networkを提案。

<img width="1014" alt="2019-02-25 18 18 06" src="https://user-images.githubusercontent.com/1038072/53326783-bdaa2e80-3929-11e9-8f74-eb369e341707.png">

## 新規性
論理的推論に適したMACセルを提案している。MACセルはLSTMセルなどと似た構造になっているが、MemoryとControlを明示的に分離している点が異なる。
そして、MACセルを使用した、MAC Netoworkを提案している。

CLEVERタスクにおいて、数え上げ（Counting）と数値的な比較（Numerical Comparison）について、既存の手法を圧倒している。

### MAC Network
MAC Networkは以下の3つのユニットからなる。

<img width="775" alt="2019-03-01 11 54 28" src="https://user-images.githubusercontent.com/1038072/53613240-cd22c380-3c18-11e9-812f-10f7a6b31dea.png">

#### Input Unit
* question: 自然言語による質問文を双方向LSTMによって学習済みWord2Vec表現の埋め込み系列cwに変換され、question表現qとして双方向の最終層を用いる。のちに、position-aware vector q_iが、その線形変換として、推論ステップの各位置で求められる。
* image: 画像から学習済みResNet 101のconv4層によって抽出された情報をkbとして獲得する。
#### MAC Recurrent Network
p個のMACセルを再帰的に繋いだ部分。セル内の各所でqやkbを受け取り、control層とmemory層を再帰的に計算する。最終層のmemory出力を最終的な出力とする。
### MAC Cell
MACセルは、以下の3つの部分からなる。
文章での説明より図を見ればいいので、各ユニットの詳細は省略する。なお、最初のステップに入力されるc_0とm_0は、それぞれ学習対象のパラメータである。

<img width="651" alt="2019-03-01 11 55 08" src="https://user-images.githubusercontent.com/1038072/53613262-e1ff5700-3c18-11e9-9453-5489bd26dfd8.png">

#### Control Unit
<img width="618" alt="control" src="https://user-images.githubusercontent.com/1038072/53613361-3c98b300-3c19-11e9-9f89-4749a42676eb.png">

ソフトAttentionを用いることで、MACセルが質問文のどこを見ているかを解釈できる。

#### Read Unit
<img width="637" alt="read" src="https://user-images.githubusercontent.com/1038072/53613366-3e627680-3c19-11e9-8c5c-ee37a198c85b.png">

KBから情報を抽出する際に、Attention Mapを作成する。Control Unitで得られたしたAttentionとRead Unitで得られたAttention Mapを並べてみると、質問文と画像からの情報抽出が対応づけられていることがわかる。

<img width="345" alt="2019-03-01 12 17 04" src="https://user-images.githubusercontent.com/1038072/53614095-f2650100-3c1b-11e9-9dec-291c4dc31c03.png">

#### Write Unit
<img width="632" alt="write" src="https://user-images.githubusercontent.com/1038072/53613374-428e9400-3c19-11e9-8887-71d37d22bcac.png">

次のMemory状態を計算する。

## 実験
視覚的推論を行うCLEVERタスクで評価。具体的には以下のようなタスクが含まれている。
* 推移関係
* 数え上げ
* 物体の性質の比較

### 使用データ
* A training set of 70,000 images and 699,989 questions
* A validation set of 15,000 images and 149,991 questions
* A test set of 15,000 images and 14,988 questions

詳細は[CLEVERのホームページ](https://cs.stanford.edu/people/jcjohns/clevr/)を参照。
<img width="513" alt="2019-03-01 12 32 47" src="https://user-images.githubusercontent.com/1038072/53614813-86d06300-3c1e-11e9-8e00-f4f815de8764.png">

また、CLEVER-Humansという人によって作られた質問文のデータセットがあり、そちらでも評価している。

### 結果
既存手法よりも良い精度を出せていることよりも、学習の効率の方が重要。
Titan X1台で15-30時間ほどで学習が完了している（10-20エポック）。

間違いとしては、オクルージョンと境界値で生じる数え上げのミスが多い。離散的な数え上げよりも、連続値による数え上げの方が良い結果になるかもしれない。

CLEVR-Humansに関しては、間違いのパターンは多岐にわたる。間違いの半数はCLEVRの教師データに含まれないタイプの質問で、物理的な属性（照明と影、反射、物体の安定性など）、相対的な距離、相対的な量（物体のうちの半数、など）、否定疑問文を含む質問が挙げられる。

## 議論
実世界でのVQAへの適用が期待できる。

## コメント
自然言語の処理の部分（Control Unit）を画像認識部分（Read Unit）の上位におき、分離しているのが面白い。

CLEVRデータセットは、自分で作ることも可能。[ソースコードが公開](https://github.com/facebookresearch/clevr-dataset-gen)されている。

アーキテクチャはそこまで難しくなく、学習時間も現実的な範囲なので実装してみたい。
